controllers/s3instance_controller.go

```diff
@@ -0,0 +1,358 @@
+		// with the call to AddFinalizer/RemoveFinalizer), and worse, causes "freshness" problem (with the
+		// "the object has been modified; please apply your changes to the latest version and try again" error)
+		err := r.Update(ctx, s3InstanceResource)
+		if err != nil {
+			logger.Error(err, "Failed to remove finalizer.")
+			return r.setS3InstanceStatusConditionAndUpdate(ctx, s3InstanceResource, "OperatorFailed", metav1.ConditionFalse, "S3InstanceFinalizerRemovalFailed",
+				fmt.Sprintf("An error occurred when attempting to remove the finalizer from s3Instance %s", s3InstanceResource.Name), err)
+		}
+	}
+	return ctrl.Result{}, nil
+}
+
+// SetupWithManager sets up the controller with the Manager.*
+func (r *S3InstanceReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	// filterLogger := ctrl.Log.WithName("filterEvt")
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&s3v1alpha1.S3Instance{}).
+		// See : https://sdk.operatorframework.io/docs/building-operators/golang/references/event-filtering/
+		WithEventFilter(predicate.Funcs{
+			// Ignore updates to CR status in which case metadata.Generation does not change,
+			// unless it is a change to the underlying Secret
+			UpdateFunc: func(e event.UpdateEvent) bool {
+				return e.ObjectOld.GetGeneration() != e.ObjectNew.GetGeneration()
+			},
+			DeleteFunc: func(e event.DeleteEvent) bool {
+				// Evaluates to false if the object has been confirmed deleted.
+				return !e.DeleteStateUnknown
+			},
+		}).
+		WithOptions(controller.Options{MaxConcurrentReconciles: 10}).
+		Complete(r)
+}
+
+func (r *S3InstanceReconciler) setS3InstanceStatusConditionAndUpdate(ctx context.Context, s3InstanceResource *s3v1alpha1.S3Instance, conditionType string, status metav1.ConditionStatus, reason string, message string, srcError error) (ctrl.Result, error) {
+	logger := log.FromContext(ctx)
+
+	// We moved away from meta.SetStatusCondition, as the implementation did not allow for updating
+	// lastTransitionTime if a Condition (as identified by Reason instead of Type) was previously
+	// obtained and updated to again.
+	s3InstanceResource.Status.Conditions = utils.UpdateConditions(s3InstanceResource.Status.Conditions, metav1.Condition{
+		Type:               conditionType,
+		Status:             status,
+		Reason:             reason,
+		LastTransitionTime: metav1.NewTime(time.Now()),
+		Message:            message,
+		ObservedGeneration: s3InstanceResource.GetGeneration(),
+	})
+
+	err := r.Status().Update(ctx, s3InstanceResource)
+	if err != nil {
+		logger.Error(err, "an error occurred while updating the status of the S3Instance resource")
+		return ctrl.Result{}, utilerrors.NewAggregate([]error{err, srcError})
+	}
+	return ctrl.Result{}, srcError
+}
+
+func (r *S3InstanceReconciler) finalizeS3Instance(ctx context.Context, s3InstanceResource *s3v1alpha1.S3Instance) error {
+	logger := log.FromContext(ctx)
+	// Create S3Client
```

Comment:
ensure no resource references this instance before deleting it.

Cf. https://gitlab.com/kubitus-project/grafana-org-operator/-/blob/dc659439102018722ed5f757554f164100af2264/internal/controller/grafanainstance_controller.go#L173-197